pandas/tests/io/formats/test_format.py

```diff
@@ -3240,14 +3240,29 @@ def test_repr_html_ipython_config(ip):
 
 
 @pytest.mark.parametrize("method", ["to_string", "to_html", "to_latex"])
+@pytest.mark.parametrize("encoding", [None, "utf-8", "gbk", "foo"])
 def test_filepath_or_buffer_arg(
-    float_frame, method, filepath_or_buffer, assert_filepath_or_buffer_equals
+    float_frame,
+    method,
+    filepath_or_buffer,
+    assert_filepath_or_buffer_equals,
+    encoding,
+    filepath_or_buffer_id,
 ):
     df = float_frame
-    expected = getattr(df, method)()
 
-    getattr(df, method)(buf=filepath_or_buffer)
-    assert_filepath_or_buffer_equals(expected)
+    if filepath_or_buffer_id not in ["string", "pathlike"] and encoding is not None:
```

Comment:
Can you split the expectation of raising into a different test called `test_filepath_or_buffer_arg_raises`? Easier convention to follow

Reply:
Do you mean I have a separate function called `test_filepath_or_buffer_arg` that tests if `to_string`, `to_latex`, `to_html` work with different `file_path_or_buffer`, and a separate `test_filepath_or_buffer_arg_raises` that also takes into account the encoding?

Reply:
Hmm I guess I'm just struggling on readability of this test. Is this generating a Cartesian product with the `filepath_or_buffer` fixture or is it zipped up as an argument because it has the same length?

Reply:
It is generating a Cartesian product.

Reply:
@simonjayhawkins 
https://github.com/pandas-dev/pandas/blob/835cdb8f81dea31d2c41eeac371ec1d519f2fb91/pandas/tests/io/formats/test_format.py#L3244-L3246

generates Cartesian product.

Reply:
The behaviour depends on the combination of `buf` and `encoding`, so I thought best to have a single test with the permutations of `buf` and `encoding`.

IMO separate tests could make this less clear, or leave gaps.

Reply:
does a refactor make it clearer?

```diff
diff --git a/pandas/tests/io/formats/test_format.py b/pandas/tests/io/formats/test_format.py
index f48d55a05..300edc3e2 100644
--- a/pandas/tests/io/formats/test_format.py
+++ b/pandas/tests/io/formats/test_format.py
@@ -3,6 +3,7 @@ Test output formatting for Series/DataFrame, including to_string & reprs
 """
 
 from datetime import datetime
+from functools import partial
 from io import StringIO
 import itertools
 from operator import methodcaller
@@ -3254,17 +3255,20 @@ def test_filepath_or_buffer_arg(
     df = float_frame
     if encoding == "gbk":
         float_frame.iloc[0, 0] = "造成输出中文显示乱码"
+
+    do_test = partial(getattr(df, method), buf=filepath_or_buffer, encoding=encoding)
+
     if filepath_or_buffer_id not in ["string", "pathlike"] and encoding is not None:
         with pytest.raises(
             ValueError, match="buf is not a file name and encoding is specified."
         ):
-            getattr(df, method)(buf=filepath_or_buffer, encoding=encoding)
+            do_test()
     elif encoding == "foo":
         with pytest.raises(LookupError, match="unknown encoding"):
-            getattr(df, method)(buf=filepath_or_buffer, encoding=encoding)
+            do_test()
     else:
         expected = getattr(df, method)()
-        getattr(df, method)(buf=filepath_or_buffer, encoding=encoding)
+        do_test()
         assert_filepath_or_buffer_equals(expected)
 
```

Reply:
Oh OK thanks for clarifying. Nice idea but I don't think the partial would change that much, so OK as is (though if you want the partial change feel free as well)